notes/trees.md

```diff
@@ -0,0 +1,234 @@
+Term.While            | `while(p){ d }`
+Term.Do               | `do { d } while(p)`
+Term.For              | `for{x <- xs} f(x)`
+Term.ForYield         | `for{x <- xs} yield f(x)`
+Term.New              | `new A`
+Term.NewAnonymous     | `new A { }`
+Term.Placeholder      | `_`
+Term.Eta              | `foo _`
+Term.Repeated         | `foo(x: _*)`
+Term.Param            | `def foo(x: Int)`
+Term.Interpolate      | `s"foo ~ $bar ~ ${42}"`
+Term.Xml              | `<h1>Hello ~ {bar} ~ {42}</h1>`
+
+## Types (`meta.Type`)
+
+```scala
+"A with B".parse[Type].get.structure
+```
+
+Ref            | Examples
+--------------------------
+Type.Name      | `B`
+Type.Select    | `a.B`
+Type.Project   | `a#B`
+Type.Singleton | `t.type`
+
+Type                  | Examples
+----------------------|--------------------
+Type.Apply            | `F[T]`
+Type.ApplyInfix       | `K Map V`
+Type.Function         | `A => B`
+Type.ImplicitFunction | `implicit A => B`
+Type.Tuple            | `(A, B)`
+Type.With             | `A with B`
+Type.And              | `A & B`
+Type.Or               | `A \| B`
+Type.Refine           | `A { def f: Int }`
+Type.Existential      | `A forSome { type T }`
+Type.Annotate         | `T @A`
+Type.Lambda           | `[X] => (X, X)`
+Type.Method           | see bellow
+Type.Placeholder      | `T[_ >: Nothing <: Any]`
+Type.Bounds           | `def F[T >: Nothing <: Any] = 1`
+Type.ByName           | `def f(v: => T) = v`
+Type.Repeated         | `def print(v: Any*): Unit`
+Type.Var              | `case _: List[t] =>`
+Type.Param            | `trait A[X]`
+
+`Type.Method`
+
+https://www.scala-lang.org/files/archive/spec/2.12/03-types.html#method-types
+
+```scala
+implicit val d = dialects.Scala212.copy(allowMethodTypes = true)
+"(e: T): T".parse[Type].get
+
+Type.Method(
+  List(
+    List(
+      Term.Param(
+        Nil,
+        Term.Name("e"),
+        Some(
+          Type.Name("T")
+        ),
+        None
+      )
+    )
+  ),
+  Type.Name("T")
+)
+```
+
+
+## Patterns (meta.Pat) and Cases (meta.Case)
+
+```scala
+"""_: Foo | _: Bar""".parse[Pat].get.structure
+```
+
+Pat                | Examples
+-------------------|----------------------------
+Lit                | `1`
+Pat.Wildcard       | `_`
+Pat.SeqWildcard    | `_*`
+Pat.Var            | `foo`
+Pat.Bind           | `foo @ Ext`
+Pat.Alternative    | `1 | 2`
+Pat.Tuple          | `(foo, bar)`
+Pat.Extract        | `Ext(foo, bar)`
+Pat.ExtractInfix   | `foo Ext bar`
+Pat.Interpolate    | `r"example (.+)$foo"`
+Pat.Xml            | `<h1>Hello, World!</h1>`
+Pat.Typed          | `foo: Int`
+Term.Name          | `foo`
+Term.Select        | `foo.bar`
+
+
+```scala
+"""case a => ()""".parse[Case].get.structure
+```
+
+Case               | Examples
+-------------------|----------------------------
+Case               | `p"case $pat if $expropt => $expr"`
+
+## Name (`meta.Name`)
+
+`meta.name.Anonymous()`
+
+```scala
+"class B { }".parse[Source].get.structure
+
+Defn.Class(
```

Comment:
The `width=1` formatting is not helping readability in my opinion, something like width=40 or 60 might be better.

Reply:
Might be good to include label names as well.